sdks/python/apache_beam/runners/interactive/extensions/apache-beam-jupyterlab-sidepanel/src/kernel/KernelModel.ts

```diff
@@ -0,0 +1,161 @@
+// distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+/**
+ * The module holds the model that handles messaging between the frontend and
+ * the connected kernel.
+ */
+
+import { ISessionContext } from '@jupyterlab/apputils';
+
+import {
+  IDisplayData,
+  IDisplayUpdate,
+  IExecuteResult
+} from '@jupyterlab/nbformat';
+
+import { Kernel, KernelMessage } from '@jupyterlab/services';
+
+import { ISignal, Signal } from '@lumino/signaling';
+
+import { KernelCode } from '../kernel/KernelCode';
+
+export class KernelModel {
+  constructor(sessionContext: ISessionContext, enableConsoleLog = false) {
+    this._sessionContext = sessionContext;
+    this._enableConsoleLog = enableConsoleLog;
+  }
+
+  get future(): Kernel.IFuture<
+    KernelMessage.IExecuteRequestMsg,
+    KernelMessage.IExecuteReplyMsg
+  > | null {
+    return this._future;
+  }
+
+  set future(
+    value: Kernel.IFuture<
+      KernelMessage.IExecuteRequestMsg,
+      KernelMessage.IExecuteReplyMsg
+    > | null
+  ) {
+    if (this._future === value) {
+      return;
+    }
+
+    if (this._future) {
+      this._future.dispose();
+    }
+
+    this._future = value;
+
+    if (!value) {
+      return;
+    }
+
+    value.onIOPub = this._onIOPub.bind(this);
+  }
+
+  get executeResult(): object {
+    if (this._executeResult) {
+      const dataInPlainText = this._executeResult.data['text/plain'] as string;
+      if (dataInPlainText) {
+        try {
+          // The slice removes trailing single quotes from the nbformat output.
+          // The replace removes literal backslashes from the nbformat output.
+          const dataInJsonString = dataInPlainText
+            .slice(1, -1)
+            .replace(/\\'/g, "'");
+          return JSON.parse(dataInJsonString);
+        } catch (e) {
+          console.error(e);
+          return {};
+        }
+      }
+    }
+    return {};
+  }
+
+  get displayData(): Array<IDisplayData> {
+    return this._displayData;
+  }
+
+  get displayUpdate(): Array<IDisplayUpdate> {
+    return this._displayUpdate;
+  }
+
+  get stateChanged(): ISignal<KernelModel, void> {
+    return this._stateChanged;
+  }
+
+  execute(code: string, expectReply = true): void {
+    // Dispose the kernel future so that no more IOPub will be handled.
+    if (this.future) {
+      this.future.dispose();
+      this.future = null;
+    }
+    // Clear the outputs from previous kernel executions.
+    this._executeResult = null;
+    this._displayData.length = 0;
+    this._displayUpdate.length = 0;
+    if (!this._sessionContext || !this._sessionContext.session?.kernel) {
+      return;
+    }
+    this.future = this._sessionContext.session?.kernel?.requestExecute({
+      code: KernelCode.COMMON_KERNEL_IMPORTS + code,
+      silent: !expectReply,
```

Comment:
If silent is the name of the underlying api, why do we use an oppositely behaving `expectReply` flag?

Reply:
The problem with `silent` is that it's confusing with the existence of `store_history`.

Intuitively, if `silent`, the kernel execution should not store history. However, it's not that case.

`silent` only means if the execution should return a result. The history is stored. The side effects are obvious.

So you will see execution_count increased even when you set `silent` to true.

Facading it with `expectReply` makes it clear when using the model.